crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -1,260 +1,524 @@
-    world::World,
+    schedule::{ShouldRun, SystemSet},
+    system::{In, IntoChainSystem, IntoSystem, Local, Res, ResMut, System},
 };
-use bevy_utils::HashMap;
-use std::{mem::Discriminant, ops::Deref};
 use thiserror::Error;
 
-pub(crate) struct StateStages {
-    update: Box<dyn Stage>,
-    enter: Box<dyn Stage>,
-    exit: Box<dyn Stage>,
-}
-
-impl Default for StateStages {
-    fn default() -> Self {
-        Self {
-            enter: Box::new(SystemStage::parallel()),
-            update: Box::new(SystemStage::parallel()),
-            exit: Box::new(SystemStage::parallel()),
-        }
-    }
+/// ### Stack based state machine
+///
+/// This state machine has three operations: Push, Pop, and Next.
+/// * Push pushes a new state to the state stack, pausing the previous state
+/// * Pop removes the current state, and unpauses the last paused state.
+/// * Next unwinds the state stack, and replaces the entire stack with a single new state
+#[derive(Debug)]
+pub struct State<T: Component + Clone + Eq> {
+    transition: Option<StateTransition<T>>,
+    stack: Vec<T>,
+    scheduled: Option<ScheduledOperation<T>>,
+    ran_update: bool,
 }
 
-pub struct StateStage<T> {
-    stages: HashMap<Discriminant<T>, StateStages>,
+#[derive(Debug)]
+enum StateTransition<T: Component + Clone + Eq> {
+    PreStartup,
+    Startup,
+    // The parameter order is always (leaving, entering)
+    ExitingToResume(T, T),
+    ExitingFull(T, T),
+    Entering(T, T),
+    Resuming(T, T),
+    Pausing(T, T),
 }
 
-impl<T> Default for StateStage<T> {
-    fn default() -> Self {
-        Self {
-            stages: Default::default(),
-        }
-    }
+#[derive(Debug)]
+enum ScheduledOperation<T: Component + Clone + Eq> {
+    Next(T),
+    Pop,
+    Push(T),
 }
 
-#[allow(clippy::mem_discriminant_non_enum)]
-impl<T> StateStage<T> {
-    pub fn with_enter_stage<S: Stage>(mut self, state: T, stage: S) -> Self {
-        self.set_enter_stage(state, stage);
-        self
-    }
-
-    pub fn with_exit_stage<S: Stage>(mut self, state: T, stage: S) -> Self {
-        self.set_exit_stage(state, stage);
-        self
-    }
-
-    pub fn with_update_stage<S: Stage>(mut self, state: T, stage: S) -> Self {
-        self.set_update_stage(state, stage);
-        self
+impl<T: Component + Clone + Eq> State<T> {
+    pub fn on_update(s: T) -> impl System<In = (), Out = ShouldRun> {
+        (|state: Res<State<T>>, pred: Local<Option<T>>| {
+            state.stack.last().unwrap() == pred.as_ref().unwrap() && state.transition.is_none()
+        })
+        .system()
+        .config(|(_, pred)| *pred = Some(Some(s)))
+        .chain(should_run_adapter::<T>.system())
     }
 
-    pub fn set_enter_stage<S: Stage>(&mut self, state: T, stage: S) -> &mut Self {
-        let stages = self.state_stages(state);
-        stages.enter = Box::new(stage);
-        self
+    pub fn on_inactive_update(s: T) -> impl System<In = (), Out = ShouldRun> {
```

Comment:
Could we have a doc string describing this function? It's not clear from either the name or a quick look at the code.